libraries/AC_WPNav/AC_WPNav.cpp

```diff
@@ -374,158 +448,134 @@ void AC_WPNav::get_wp_stopping_point(Vector3f& stopping_point) const
+        track_error = _pos_control.get_pos_error().dot(track_direction);
+        track_velocity = _inav.get_velocity().dot(track_direction);
+        // set time scaler to be consistent with the achievable aircraft speed with a 5% buffer for short term variation.
+        track_scaler_dt = constrain_float(0.05f + (track_velocity - _pos_control.get_pos_xy_p().kP() * track_error) / target_velocity.length(), 0.1f, 1.0f);
+        // set time scaler to not exceed the maximum vertical velocity during terrain following.
+        if (is_positive(target_velocity.z)) {
+            track_scaler_dt = MIN(track_scaler_dt, fabsf(_wp_speed_up_cms / target_velocity.z));
+        } else if (is_negative(target_velocity.z)) {
+            track_scaler_dt = MIN(track_scaler_dt, fabsf(_wp_speed_down_cms / target_velocity.z));
         }
+    } else {
+        track_scaler_dt = 1.0f;
     }
-    // advance the current target
-    if (!reached_leash_limit) {
-    	_track_desired += _limited_speed_xy_cms * dt;
-
-    	// reduce speed if we reach end of leash
-        if (_track_desired > track_desired_max) {
-        	_track_desired = track_desired_max;
-        	_limited_speed_xy_cms -= 2.0f * _track_accel * dt;
-        	if (_limited_speed_xy_cms < 0.0f) {
-        	    _limited_speed_xy_cms = 0.0f;
-        	}
-    	}
+    // change s-curve time speed with a time constant of maximum acceleration / maximum jerk
+    float track_scaler_tc = 0.01f * _wp_accel_cmss/_wp_jerk;
+    if (!is_zero(_wp_jerk)) {
+        track_scaler_tc = 0.01f * _wp_accel_cmss/_wp_jerk;
+    } else {
+        track_scaler_tc = 1.0f;
     }
+    _track_scalar_dt += (track_scaler_dt - _track_scalar_dt) * (dt / track_scaler_tc);
 
-    // do not let desired point go past the end of the track unless it's a fast waypoint
-    if (!_flags.fast_waypoint) {
-        _track_desired = constrain_float(_track_desired, 0, _track_length);
+    if (!_this_leg_is_spline) {
+        // update target position, velocity and acceleration
+        target_pos = _origin;
+        s_finished = _scurve_this_leg.advance_target_along_track(_scurve_prev_leg, _scurve_next_leg, _wp_radius_cm, _flags.fast_waypoint, _track_scalar_dt * dt, target_pos, target_vel, target_accel);
     } else {
-        _track_desired = constrain_float(_track_desired, 0, _track_length + WPNAV_WP_FAST_OVERSHOOT_MAX);
+        // splinetarget_vel
+        target_vel = target_velocity;
+        _spline_this_leg.advance_target_along_track(_track_scalar_dt * dt, target_pos, target_vel);
+        s_finished = _spline_this_leg.reached_destination();
     }
 
-    // recalculate the desired position
-    Vector3f final_target = _origin + _pos_delta_unit * _track_desired;
+    // input shape the terrain offset
+    shape_pos_vel(terr_offset, 0.0f, _pos_terrain_offset, _vel_terrain_offset, _accel_terrain_offset, 0.0f, 0.0f, _pos_control.get_max_accel_z()*4.0f, 0.05f, _track_scalar_dt * dt);
```

Comment:
`*4.0f`?

Reply:
Hmmmmmm I don't remember that. It maybe there for doing some testing or something. I would suggest we remove it.

Reply:
removed the *4.0f.